rs/nns/governance/src/network_economics.rs

```diff
@@ -0,0 +1,263 @@
+        let from_range = {
+            let begin = self.get_start_reducing_voting_power_after_seconds();
+            let end = begin + self.get_clear_following_after_seconds();
+
+            begin..end
+        };
+
+        #[allow(clippy::reversed_empty_ranges)]
+        let to_range = 1..0;
+
+        LinearMap::new(from_range, to_range)
+    }
+
+    pub fn get_start_reducing_voting_power_after_seconds(&self) -> u64 {
+        self.start_reducing_voting_power_after_seconds
+            .unwrap_or(Self::DEFAULT_START_REDUCING_VOTING_POWER_AFTER_SECONDS)
+    }
+
+    pub fn get_clear_following_after_seconds(&self) -> u64 {
+        self.clear_following_after_seconds
+            .unwrap_or(Self::DEFAULT_CLEAR_FOLLOWING_AFTER_SECONDS)
+    }
+}
+
+// Seems like there is probably some way we can make #[derive(...)] generate
+// implementations of this, because the hand-crafted implementations below are
+// pretty dang (if not completely) repetitive. OTOH, it would be a lot of work,
+// and it is not clear that we would use it many times.
+pub(crate) trait InheritFrom {
+    /// Returns a modified copy of self where fields containing 0 are replaced
+    /// with the value from base.
+    fn inherit_from(&self, base: &Self) -> Self;
+}
+
+// Ideally, we'd use num_traits::Zero to give a generic implementation that
+// applies to all integer types, but Rust refuses to allow that in the presence
+// of impl InheritFrom for Option<T> below. If only there was some way we could
+// tell Rust, "ok, but this impl does not apply when the type happens to be
+// Option", but that doesn't exist (yet). Fortunately, we do not use a wide
+// range of integer types. Therefore, only this one is needed (for now).
+impl InheritFrom for u64 {
+    fn inherit_from(&self, base: &Self) -> Self {
+        if self == &0_u64 {
+            return *base;
+        }
+
+        *self
+    }
+}
+
+impl InheritFrom for u32 {
+    fn inherit_from(&self, base: &Self) -> Self {
+        if self == &0_u32 {
+            return *base;
+        }
+
+        *self
+    }
+}
+
+impl InheritFrom for ic_nervous_system_proto::pb::v1::Decimal {
+    fn inherit_from(&self, base: &Self) -> Self {
+        if self
+            == &(ic_nervous_system_proto::pb::v1::Decimal {
+                human_readable: Some("0".to_string()),
+            })
+        {
+            return base.clone();
+        }
+
+        self.clone()
+    }
+}
+
+impl InheritFrom for Percentage {
+    fn inherit_from(&self, base: &Self) -> Self {
+        if self
+            == &(Percentage {
+                basis_points: Some(0),
+            })
+        {
+            return *base;
+        }
+
+        *self
+    }
+}
+
+impl<T> InheritFrom for Option<T>
+where
+    T: InheritFrom + Clone,
+{
+    fn inherit_from(&self, base: &Self) -> Self {
+        match (self, base) {
+            (Some(me), Some(base)) => Some(me.inherit_from(base)),
```

Comment:
Do I get it rightly that this code does the following?

```
let payload = Payload {
   foo: Some(42_u64),
};

let new_payload = payload.inherit_from(Payload {
    foo: Some(0_u64)
})

assert_eq!(new_payload, Payload {
    foo: Some(42_u64)
})
```

because both `self` and `base` are `Some`, so the inheritance takes place from the rules for `u64`?

This seems sub-optimal, as if the API payload has a field `foo: Option<u64>`, the way to express that we don't want to change that is to write `foo: None`, not `foo: Some(0)`.

Reply:
> so the inheritance takes place from the rules for `u64`?

Yes.

> `foo: None`

Yes, this is how "do not touch foo" is expressed.

I think where we might be confused is that the order of arguments is like this:

```
from_the_proposal.inherit_from(original)
```

Then, if `from_the_proposal` has `foo: None`, then `original` clobbers `foo`.

Reply:
Ah okay, thanks for clarifying!